src/persist/src/indexed/mod.rs

```diff
@@ -858,14 +858,15 @@ impl<L: Log, B: Blob> Indexed<L, B> {
             .push(PendingResponse::Unit(res, resp));
     }
 
-    fn do_allow_compaction(&mut self, id: Id, since: u64) -> Result<(), Error> {
-        let trace = self
-            .traces
-            .get_mut(&id)
-            .ok_or_else(|| Error::from(format!("never registered: {:?}", id)))?;
-        let since = Antichain::from_elem(since);
+    fn do_allow_compaction(&mut self, id_sinces: Vec<(Id, Antichain<u64>)>) -> Result<(), Error> {
+        for (id, since) in id_sinces {
```

Comment:
i think you need two separate loops here, one to check that all the compactions are valid, and then another to do them, as otherwise we might leave the metadata in an incorrect state

Reply:
Makes sense when I think about it real hard, but still feels like a footgun. I bet there's something better we can do here.

Anyway, done!